src/lib/cache_lib/intf.ml

```diff
@@ -0,0 +1,111 @@
+open Async_kernel
+open Core_kernel
+
+module Constant = struct
+  module type S = sig
+    type t
+
+    val t : t
+  end
+end
+
+module Cached = struct
+  module type S = sig
+    type ('t, 'cache_t) t
+
+    val phantom : 't -> ('t, _) t
+
+    val peek : ('t, _) t -> 't
+
+    val invalidate : ('t, _) t -> 't Or_error.t
+
+    val was_consumed : (_, _) t -> bool
+
+    val transform : ('t0, 'cache_t) t -> f:('t0 -> 't1) -> ('t1, 'cache_t) t
+    (** [transform] maps a [Cached.t], consuming the original in the process *)
+
+    val lift_deferred :
+      ('t Deferred.t, 'cache_t) t -> ('t, 'cache_t) t Deferred.t
+
+    val lift_result :
+      (('t, 'e) Result.t, 'cache_t) t -> (('t, 'cache_t) t, 'e) Result.t
```

Comment:
these are typically called `sequence` (so `sequence_result` and `sequence_deferred`). see https://github.com/purescript/purescript-foldable-traversable/blob/master/src/Data/Traversable.purs#L52 (this operation is also called sequence in Haskell, Scala, Kotlin, Swift, and F# communities)

Reply:
That word feels wrong to me, but I'll trust you on this one.